python_modules/dagster/dagster/core/object_store.py

```diff
@@ -27,8 +28,46 @@ def ensure_boto_requirements():
     return (boto3, botocore)
 
 
-@six.add_metaclass(ABCMeta)
-class ObjectStore:
+class TypeStoragePlugin(six.with_metaclass(ABCMeta)):  # pylint: disable=no-init
+    '''Base class for storage plugins.
+
+    Extend this class for (storage_mode, runtime_type) pairs that need special handling.
+    '''
+
+    @classmethod
+    @abstractmethod
+    def set_object(cls, object_store, obj, context, runtime_type, paths):
+        check.subclass_param(object_store, 'object_store', ObjectStore)
+        return object_store.set_object(obj, context, runtime_type, paths)
+
+    @classmethod
+    @abstractmethod
+    def get_object(cls, object_store, context, runtime_type, paths):
+        check.subclass_param(object_store, 'object_store', ObjectStore)
+        return object_store.get_object(context, runtime_type, paths)
+
+
+class ObjectStore(six.with_metaclass(ABCMeta)):
+    _override_methods = ['set_object', 'get_object', 'has_object', 'rm_object']
+
+    def __init__(self, types_to_register=None):
+        types_to_register = check.opt_dict_param(
+            types_to_register,
+            'types_to_register',
+            key_type=RuntimeType,
```

Comment:
I think indexing against type name (a string) will be more safe and accurate here.